### ISSUE TITLE
StackSlot identifier variables are not removed after PrettifyAssignments

### ISSUE THREAD

**Issue by Chicken-Bones** [icsharpcode/ILSpy#1126]
Test Case
```cs
public struct Point
{
	public int x, y;
	
	public static Point operator+(Point p, int i) => new Point();
	public static Point operator+(Point p1, Point p2) => new Point();
}

public class Class
{
	private int[] iArr;
	private Point[] pArr;
	private Point p;
	
	private Class Inst => null;

	public void Inc() {
		iArr[0] += 1;
		pArr[0] += 1;

		pArr[0].x += 1;
		pArr[0] += pArr[0];
		
		Inst.p += 1;
		Inst.p.x += 1;
	}
}
```

Decompiled
```cs
public void Inc()
{
	iArr[0]++;
	ref Point val = ref pArr[0];
	val += 1;
	pArr[0].x++;
	ref Point val2 = ref pArr[0];
	val2 += pArr[0];
	Inst.p.x++;
	Class inst = Inst;
	inst.p += 1;
}
```

**Comment by siegfriedpammer** [comment]
Modifying variables, specially those with address usage (ref variables are address usage), should only be done in the IL Ast, because there we have the appropriate infrastructure and APIs to make larger-scale changes to the code and ensure its semantic correctness. In addition, if anything goes horribly wrong you will trigger tons of assertions, that ensure the correctness of the whole tree.

The `PrettifyAssignments` step should only transform "x = x op y" into "x op= y" where possible, i.e. change a single C# expression. This is necessary, because custom operators are not yet transformed from method calls to operators in the IL Ast. In the ILAst these are represented as method calls.

Please take a look at `ICSharpCode.Decompiler.IL.Transforms.TransformAssignment` for the transforms and the decompiler pipeline in `CSharpDecompiler` (see GetILTransforms).

Unfortunately, there is only little documentation on the IL Ast, so best is to use the "debug steps" tree view, which allows you to analyze each IL Ast transformation step / loop.

It can be activated by launching a debug build of ILSpy:
![grafik](https://user-images.githubusercontent.com/344204/39288094-ee5efcde-4926-11e8-9204-4172f432c264.png)

1) Select a single method in the tree view. (Does not work with other members or types)
2) Switch to "ILAst" in the languages dropdown.
3) Press the "Show Steps" button.

Now you can see a tree representing all transformations at the top:
![grafik](https://user-images.githubusercontent.com/344204/39288212-65b2746e-4927-11e8-9811-ba82565ce4c4.png)

You can manually repeat the transforms by double clicking on the steps.

Hope this helps at least a little bit to get started! If you have any other questions, feel free to ask!

Thank you for your help!

**Comment by Chicken-Bones** [comment]
I did notice the ILAst debug steps, nice UI you have there. I also found `HandleStObjCompoundAssign`

"This is necessary, because custom operators are not yet transformed from method calls to operators in the IL Ast. In the ILAst these are represented as method calls."

From what I've gleaned, the `IAstTransform`s are for the language dependent constructs, syntax sugars and things that require larger program scope to evaluate such as qualifiers and extension methods.

Are you suggesting we produce `CompoundAssignmentInstruction` for overloaded operators. This would require moving `ReplaceMethodCallsWithOperators` into the ILAst (which I don't see an immediate problem with)

**Comment by dgrunwald** [comment]
Any 'remove temporaries' step must happen in the ILAst.
Consider code like:
`BigInteger[] arr = { a, GetInstance().MyBigInt += b, c };`
If the temporary representing the `GetInstance()` result value is not already removed when the array initializer detection runs, we can't recognize the array initializer.
But then there's also code like:
`GetInstance().SomeProperty += SomeFunction(new int[] { a, b, c });`
Now the `get_SomeProperty` and `set_SomeProperty` calls are separated by the array initializer, so we can only detect a compound assignment if we have already detected the array initializer.

Many transforms depends on sub-expressions already being recognized as a single expression; but often such transforms themselves combine multiple statements into a single expression. Thus, all such transforms depend on each other, they must run in a single interleaved pass. -> [StatementTransform](https://github.com/icsharpcode/ILSpy/blob/1ab0066b1c81abf14e2a6948038cd243b49b36c7/ICSharpCode.Decompiler/CSharp/CSharpDecompiler.cs#L121)

So that's where compound assignments with user-defined operators need to be detected.
Now, I don't think it makes sense to move all of `ReplaceMethodCallsWithOperators` into the ILAst -- e.g. the ILAst `BinaryNumericInstruction` is designed for built-in operators, re-using it for user-defined operators doesn't make much sense. And the semantics of a normal user-defined operator are already perfectly captured by a normal `call` instructions. But compound assignment needs to be special cased in the ILAst. So maybe we need a new `compound.call` instruction in the ILAst (for use only with user-defined operators). Or we change the existing `CompoundAssignmentInstruction` to also support calls in addition to the builtin operators.

**Comment by dgrunwald** [comment]
We've added ILAst support for compound assignments with user-defined operators (#1129). Your code should now decompile correctly.

**Comment by Chicken-Bones** [comment]
Thankyou for this, I was going to take a shot at implementing it in similar fashion but had a sudden lack of free time. Will this be merged into 3.1.x?

**Comment by dgrunwald** [comment]
We changed release plans. There likely won't be another 3.1.x; but we'll release a 3.2 from master before merging the ILSpy 4.0 System.Reflection.Metadata branch.
